### DISCUSSION TITLE
Self-healing and automagically-repairing Zigbee network mesh in Zigpy and ZHA?

### DISCUSSION THREAD

**Discussion by Gamester17** [zigpy/zigpy#614] [ZHA development]
How does ZHA/Zigpy handle a temporary loss of or the permanent removal of one or more Zigbee router Zigbee network mesh?

The question came up in Home Assistant's community Zigbee sub-forum https://community.home-assistant.io/t/sonoff-zbbridge-sonoff-zigbee-bridge-from-itead/187346

The question is partially about if ZHA/Zigpy and all or most standard Zigbee devices feature timed fail-over and fail-back functions for Zigbee network mesh self-healing and automatic-repairs? It is also partially about if ZHA/Zigpy can self-calibrate its Zigbee mesh network and automatically optimize it so that Zigbee end devices always try to connect to Zigbee routers with best reception for them unless you manually force a Zigbee end-device to connect to a specific Zigbee router?

Such as example try to automatically re-pair a Zigbee end device to another Zigbee router or the coordinator when the Zigbee router it is connected goes offline/disappear for a longer than normal? And if so does it then try to re-pair it with the original Zigbee router if it reappears and has been online again for a certain amount of time?

These initial questions are specific to scenarios are when Zigbee devices (Zigbee end device or other router devices) is not directly connected to the Zigbee coordinator but instead connected indirectly through a Zigbee router.

Scenario examples that are wondering about time-frames between different automatic actions that Zigpy/ZHA takes:

1. How long does it take and what exactly happens to a Zigbee end-device when the Zigbee router it is connected unexpectedly goes offline (for example due to loss of power or accidentally unplugged)? 

2. How long does it take and what exactly happens to a Zigbee end-device when the Zigbee router it was previously connected to comes back online if it has already been automatically connected to the other Zigbee router?

3. How long does it take and what exactly happens to a Zigbee end-device when the Zigbee router it is connected is manually removed in ZHA/Zigpy on purpose?

From my experience with other Zigbee hubs/gateways/bridges/controllers, (like example Samsung SmartThings Hub), I would expect the application software to after automatically self-heal the Zigbee network mesh if a Zigbee router goes offline unexpectedly and as such after while let the Zigbee end-device to connect to another Zigbee router device if available (in essence fail-over the Zigbee device to another Zigbee router). Then if the original Zigbee router comes back online it will after a while self-calibrate the Zigbee mesh network and automatically reconnect the Zigbee end-device to its original Zigbee router if it looks to be the one with the best signal reception for it.

**Comment by dmulcahey** [comment]
This is all handled by the zigbee stack for the most part and the software (meaning Zigpy) isn’t involved. SI labs has a bunch of docs that explain how this works for the ember stack (bellows). 

https://www.silabs.com/documents/public/user-guides/ug103-02-fundamentals-zigbee.pdf

Start at section 4.3

**Comment by Gamester17** [comment]
A follow-up question to this:

> unless you manually force a Zigbee end-device to connect to a specific Zigbee router?

What do you devs think about using an affinity rules system if possible for failover and failback if a Zigbee router temporarily goes offline?

If possible you could use affinity rules possibly let the end-user manually set a rule for a Zigbee device that it "must" always only connect to a specific Zigbee router even if it goes offline and other Zigbee routers are available, or the end-user could manually set a rule for a Zigbee device that it "should" connect to a specific Zigbee router, meaning that it should use primarily use that Zigbee router if available even if it sees other Zigbee routers with better reception, however, if that Zigbee router goes offline then it can use other Zigbee routers until its primary Zigbee router comes back online.

I worked a lot with VMware vSphere vCenter Server and ESX/ESXi virtualization in my profession and it has such Affinity Rules for its HA (High Availability) and DRS (Distributed Resource Scheduler) features which are its main essential core features for achieving high-availability.

I think it's interesting that it has soft affinity rules "should" follow and hard affinity rules "must" follow) for virtual machines failover/failback, noting that it also has soft and hard affinity rules for what it "must not" and "should not" do.

https://docs.vmware.com/en/VMware-vSphere/7.0/com.vmware.vsphere.resmgmt.doc/GUID-0591F865-91B5-4311-ABA6-84FBA5AAFB59.html

* Select a specification for the rule.
  * **Must run on hosts in group**. Virtual machines in VM Group 1 must run on hosts in Host Group A.
  * **Should run on hosts in group**. Virtual machines in VM Group 1 should, but are not required, to run on hosts in Host Group A.
  * **Must not run on hosts in group**. Virtual machines in VM Group 1 must never run on host in Host Group A.
  * **Should not run on hosts in group**. Virtual machines in VM Group 1 should not, but might, run on hosts in Host Group A.
Click OK.

https://docs.vmware.com/en/VMware-vSphere/7.0/com.vmware.vsphere.avail.doc/GUID-E137A9F8-17E4-4DE7-B986-94A0999CF327.html

 When you edit a DRS affinity rule, you must use vSphere HA advanced options to enforce the desired failover behaviour for vSphere HA.
 
 * **HA must respect VM anti-affinity rules during failover** -- When the advanced option for VM anti-affinity rules is set, vSphere HA does not fail over a virtual machine if doing so violates a rule. Instead, vSphere HA issues an event reporting there are insufficient resources to perform the failover.
 
 * **HA should respect VM to Host affinity rules during failover** --vSphere HA attempts to place VMs with this rule on the specified hosts if at all possible.

https://docs.vmware.com/en/VMware-vSphere/7.0/com.vmware.vsphere.resmgmt.doc/GUID-793013E2-0976-43B7-9A00-340FA76859D0.html

**Reply by walthowd** [reply]
These decisions on joining (any decisions on a desired feature like affinity) all occur on the zigbee devices themselves. Zigpy/ZHA aren't involved and can't influence the devices decision in any way.

**Comment by MattWestb** [comment]
Sleeping end device is normally the problem then some is changing parent if its cant finding it then needed prity OK and some others dot trying changing its parent and leaving the network (old Xiaomi no ZB3 sensors).

The other part is the zigbee stack in the routers if its working in newer Zigbee PRO mode on not.
All EFR32 based is PRO and is building one good multilink mesh network (both IKEA old ZLL and ZB3 is doing it well) but some other with old stacks like TI-CC253X is only making one route to the coordinator and dont open multi connections (Chinese ZB3 devices like old HOMA dimmers) but i think its if the network is formed as ZB3 or HA 1.x and some routers is working in backward compatibility mode as no sleeping end devices = not real routers but i have not finding one deep paper that explaining the backwards compatibility in zigbee 3 / PRO.

One PM from the HA postings: The mesh is rebuilding pretty fast (i think some seconds not minutes) but ZHA is not scanning the network topography very oft (some hours between is my impression) so its not one good tool for testing if not forcing ZHA doing network topography scans more often (its blocked / limetide in ZHA).

I was adding one LEDL LED stripe and was foreign ZHA scanning the network topography and in 2 minutes its have building the colmpleete mesh with multi links step by step.

I think its need other tools for testing the mesh is acting in real and also deeper knowledge for understanding how its working in real.

But if having one stabile mesh with decent routers its normally no problem losing one if its not having sleeping end devices that is being lost and the rest is working OK (my experience of ZB3 / ZB PRO routers working in ZB3 mode network).

**Reply by Gamester17** [reply]
> The other part is the zigbee stack in the routers if its working in newer Zigbee PRO mode on not.
> All EFR32 based is PRO and is building one good multilink mesh network (both IKEA old ZLL and ZB3 is doing it well) but some other with old stacks like TI-CC253X is only making one route to the coordinator and dont open multi connections (Chinese ZB3 devices like old HOMA dimmers) but i think its if the network is formed as ZB3 or HA 1.x and some routers is working in backward compatibility mode as no sleeping end devices = not real routers but i have not finding one deep paper that explaining the backwards compatibility in zigbee 3 / PRO.

Do you know if using Z-Stack 3.0.x or Z-Stack 3.x.0 on Texas Instruments based adapters with zigpy-znp always form a ZB3 network?

https://github.com/Koenkk/Z-Stack-firmware/tree/master/coordinator

Z-Stack 3.0.1 firmware on older CC2531 (and CC2530) should work with zigpy-znp but wondering if always forms a Zigbee 3.0 network?

https://github.com/zigpy/zigpy-znp

As you noted, it will of course also depends if the Zigbee router device and Zigbee end device is also a Zigbee 3.0  device or not.

**Reply by Adminiuga** [reply]
there's no difference in the "mesh" handling in ZB3 and HA1.2, in other words Zigbee Pro R21 and R22 mesh is handled the same. Most notable difference is in the device joining procedure and device/network commissioning (other less significant is child aging and device commissioning). But that is handled by the coordinator firmware and not strictly by zigpy.  
ZNP AFAIK is the only exception, cause ZNP is working at the lower abstraction level than EZSP. But you can expect that sticks with ZB3 firmware would support R22 protocol. but the mesh handling itself it won't be different.

**Reply by MattWestb** [reply]
Install ZigZag (from our Samantha) and you can see all good and bad links the routers is making / not making.
My production ZHA then starting moving devices from deCONZ in oktober (Its little larger to day and deCONZ is little slimmed :-)))
![ZigZag01](https://user-images.githubusercontent.com/49618193/103921575-f82ea600-5112-11eb-8726-40b126981cd1.PNG)
Screen is from version 1 that is now outdated but is still working until ver 2 is released.

All routers is Zigbee 3 without the CWS 600 lumen and opal 1000 its ZLL (still Zigbee PRO) and the old motion sensor.
As you can see all can rerouting good if one router is failing (except vorzimmer that have 2 meters betone direct to the coordinator one level above in diagonal need one router between for having good performance).
